placement/algo/mirrored.go

```diff
@@ -224,45 +242,68 @@ func (a mirroredAlgorithm) ReplaceInstances(
 	return p, nil
 }
 
-// allInitializing returns true when the given list of instances
-// matches all the initializing instances in the placement.
-func allInitializing(p placement.Placement, instances []string) bool {
+func (a mirroredAlgorithm) MarkShardAvailable(
+	p placement.Placement,
+	instanceID string,
+	shardID uint32,
+) (placement.Placement, error) {
+	if err := a.IsCompatibleWith(p); err != nil {
+		return nil, err
+	}
+
+	return a.shardedAlgo.MarkShardAvailable(p, instanceID, shardID)
+}
+
+// allInitializing returns true when
+// 1: the given list of instances matches all the initializing instances in the placement.
+// 2: the shards are not cutover yet.
+func allInitializing(p placement.Placement, instances []string, nowNanos int64) bool {
 	ids := make(map[string]struct{}, len(instances))
 	for _, i := range instances {
 		ids[i] = struct{}{}
 	}
 
-	return allInstancesInState(ids, p, func(i placement.Instance) bool {
-		return i.IsInitializing()
+	return allInstancesInState(ids, p, func(s shard.Shard) bool {
+		return s.State() == shard.Initializing && s.CutoverNanos() > nowNanos
 	})
 }
 
-// allInitializing returns true when the given list of instances
-// matches all the leaving instances in the placement.
-func allLeaving(p placement.Placement, instances []placement.Instance) bool {
+// allLeaving returns true when
+// 1: the given list of instances matches all the leaving instances in the placement.
+// 2: the shards are not cutoff yet.
+func allLeaving(p placement.Placement, instances []placement.Instance, nowNanos int64) bool {
 	ids := make(map[string]struct{}, len(instances))
 	for _, i := range instances {
 		ids[i.ID()] = struct{}{}
 	}
 
-	return allInstancesInState(ids, p, func(i placement.Instance) bool {
-		return i.IsLeaving()
+	return allInstancesInState(ids, p, func(s shard.Shard) bool {
+		return s.State() == shard.Leaving && s.CutoffNanos() > nowNanos
 	})
 }
 
+func instanceCheck(instance placement.Instance, shardCheckFn func(s shard.Shard) bool) bool {
+	for _, s := range instance.Shards().All() {
+		if !shardCheckFn(s) {
+			return false
+		}
+	}
+	return true
+}
+
 func allInstancesInState(
 	instanceIDs map[string]struct{},
 	p placement.Placement,
-	isInStateFn func(placement.Instance) bool,
+	forEachShardFn func(s shard.Shard) bool,
 ) bool {
-	for _, i := range p.Instances() {
-		if !isInStateFn(i) {
+	for _, instance := range p.Instances() {
+		if !instanceCheck(instance, forEachShardFn) {
 			continue
 		}
-		if _, ok := instanceIDs[i.ID()]; !ok {
+		if _, ok := instanceIDs[instance.ID()]; !ok {
 			return false
 		}
-		delete(instanceIDs, i.ID())
+		delete(instanceIDs, instance.ID())
```

Comment:
nit: perhaps it would be preferable to not modify `instanceIDs` in case we ever want to reuse it after calling `allInstancesInState`? Seems like we could get the same effect by just incrementing a counter each time through the loop instead of deleting from `instanceIDs` and the checking that the counter equals the length of `instanceIDs`?

Reply:
the instanceIDs are created purely to be used in this function, I could do the inc/dec counter for performance gain, but seems like an overkill